CLDC-1633 build feature csv download of sales logs

arfacamble opened the pull request:
DEPLOY-ACTION:
Jemma: 
_"When this hits production in my absence, could you just give Miguel, Alistair and Sam the heads up that it's there? I don't expect them to look at it straight away because they've got a publication deadline, but just so they're aware:"_
___

https://digital.dclg.gov.uk/jira/browse/CLDC-1633

One important change in the model for how this csv will be generated comes from the fact that I couldn't work out after poking around with James, how the old model for exporting lettings logs works exactly, there's some rails magic and some mysterious way of not losing the filtering passing between the Job, a LettingsLog class method and the LettingsLogCsvService.
I hope this way of doing it is easier to read, follow and test. It does mean that until the tech debt LettingsLogCsvService is tackled there are one or two places with a bit of messy looking inconsistency

arfacamble committed:
create a method on the FormHandler that returns the sales form questions for all years in the order that they appear in the form

arfacamble committed:
update csv email job to accomodate sales log export as well as lettings
add to tests to reflec the changes made

arfacamble committed:
write tests to cover the desired functionality of the SalesLogCsvService

arfacamble committed:
create the SalesLogCsvService
create a necessary method on the log to enable submission method to be included on the csv
derive values for the two halves of previous postcode for export

jamesrose commented on app/jobs/email_csv_job.rb:
```diff
@@ -5,14 +5,25 @@ class EmailCsvJob < ApplicationJob
-    filename = organisation.present? ? "logs-#{organisation.name}-#{Time.zone.now}.csv" : "logs-#{Time.zone.now}.csv"
+    filename = [log_type, "logs", organisation&.name, Time.zone.now].compact.join("-") + ".csv"
+
+    csv_string = if log_type == "sales"
```
This can be in one condition as per above

jamesrose commented on app/models/log.rb:
```diff
@@ -136,6 +136,16 @@ def field_formatted_as_currency(field_name)
+    if bulk_upload_id.present?
+      "bulk upload"
+    elsif old_id.present?
+      "migrated"
```
Do we need this one?

arfacamble committed:
add relevant links in the UI and pipe everything together in controllers
amend organisations controller to have flexibility to download logs of either type
add necessary methods to sales log controller, raising shared method to logs controller
update routing for amendments and additions
extract helper method to build urls for downloading logs within an organisation

arfacamble committed:
correct various linter complaints and tech review suggestions

arfacamble committed:
minor amendment to add old_id and reorder early columns

arfacamble committed:
undo my 'clever' refactor that broke things

arfacamble committed:
refactoring of csv service after some tech review and some UI testing in review app

arfacamble committed:
update tests to include a test of a full export and all values in teh csv

arfacamble committed:
correct minor routing error to ensure correct url is shown and tab selected after requesting csv email

arfacamble committed:
update organisations controller requests spec file to cover new functionality and make a minor amendment to authentication scope in the controller after error found in testing

arfacamble committed:
write request tests for the new functionality in the sales log controller, define authorisation in the controller

arfacamble committed:
minor correction after rubocop's kind suggestion'

arfacamble committed:
various corrections from first pass at PO, tech review, linter, etc

arfacamble committed:
refactor :ordered_sales_questions_for_all_years

arfacamble committed:
first pass at implementing flexible code-based form fixtures for testing

arfacamble commented on spec/factories/form.rb:
```diff
@@ -0,0 +1,25 @@
+    form.questions = form.pages.flat_map(&:questions)
+    form
+  end
+end
```
This is where the shape of this solution is determined.
`Form` does not have accessors like section, subsection, etc and it is initialised with an array of `Section` classes, not instances, which it initialises and from that populates various instance variables. 
So to create something that `is_a? Form` with the structure we want, I have had to create FormFixture which inherits and adds setters and recreate a section of `Form:initialize` in `FormFactory:build`

arfacamble committed:
second pass

arfacamble commented on spec/factories/page.rb:
```diff
@@ -1,5 +1,15 @@
+          page.questions = [build(:question, id: evaluator.question_id, page:)]
+        end
+      end
+    end
```
the point of making all these call downwards recursively is that on line 10 we assign the page to the question (and something similar in the other factories)
which allows code of the form [sic] `question.form` to work.
This isn't actually necessary for the test I am writing here, but I think it's an important pattern to demonstrate for this model of building forms on the fly for tests

arfacamble committed:
refactor all tests of :ordered_sales_questions_for_all_years to use new factories

arfacamble committed:
some refactoring in the testing of the csv service

asmega commented on app/controllers/logs_controller.rb:
```diff
@@ -28,6 +28,10 @@ def create
     end
   end
 
+  def codes_only_export?(params)
```
```suggestion
  def codes_only_export?
```
i think you already have access to this and therefore don't need to pass it in

arfacamble replied:
veyr good. Always nice to know where the bits of rails magic pop up regarding scope

asmega commented on app/controllers/lettings_logs_controller.rb:
```diff
@@ -124,6 +115,11 @@ def org_params
 
+  def authenticate_scope!
+    codes_only_export = codes_only_export?(params)
+    head :unauthorized and return if codes_only_export && !current_user.support?
```
```suggestion
    head :unauthorized and return if codes_only_export? && !current_user.support?
```
in tandem with my other suggestion, i feel the local variable is a bit redundant and overkill and can probably be removed

arfacamble replied:
certainly given params not needed to be passed in.

asmega commented on app/controllers/sales_logs_controller.rb:
```diff
@@ -36,11 +37,34 @@ def edit
+  end
+
+  def email_csv
+    all_orgs = params["organisation_select"] == "all" # what's this for? params['organisation_select'] appears to always be nil
```
is it one of the search filters? to restrict to all orgs, or a specific org?

arfacamble replied:
I need toi investigate this. Left it in as comment since I don't have the asnwer!

asmega commented on app/services/csv/sales_log_csv_service.rb:
```diff
@@ -0,0 +1,108 @@
+
+  private
+
+    ATTRIBUTES_OF_RELATED_OBJECTS = {
```
```
irb(main):001:0> class A
irb(main):002:1>   private
irb(main):003:1>   B = "B"
irb(main):004:1> end
=> "B"
irb(main):005:0> A::B
=> "B"
```
As an fyi declaring constants after a `private` this way does not make them private

arfacamble committed:
use that fact that params is always available in controllers and don't pass it around, inline some methods calls

arfacamble committed:
correct minor bug to ensure that "Return to logs" link returns to the correct index page

arfacamble committed:
remove reminder comments

arfacamble committed:
write further tests on the manipulation of questions into the csv headers, update factories of form constituents to allow the creation of forms with richer questions

natdeanlewissoftwire commented on app/controllers/lettings_logs_controller.rb:
```diff
@@ -86,19 +81,13 @@ def destroy
     all_orgs = params["organisation_select"] == "all"
-    codes_only_export = params.require(:codes_only) == "true"
-    EmailCsvJob.perform_later(current_user, search_term, @session_filters, all_orgs, nil, codes_only_export)
+    EmailCsvJob.perform_later(current_user, search_term, @session_filters, all_orgs, nil, codes_only_export?)
```
luscious

arfacamble replied:
that's a bit strong

natdeanlewissoftwire commented on app/controllers/organisations_controller.rb:
```diff
@@ -137,6 +139,20 @@ def sales_logs
+  def download_sales_csv
+    organisation_logs = SalesLog.visible.where(owning_organisation_id: @organisation.id)
+    unpaginated_filtered_logs = filtered_logs(organisation_logs, search_term, @session_filters)
+    codes_only = params.require(:codes_only) == "true"
```
can we use the newly moved `codes_only_export?` method for this too?

arfacamble replied:
not quite, since it's calling a method in the logs controller, but I can recreate it in private section and make looky nicer

natdeanlewissoftwire commented on app/controllers/organisations_controller.rb:
```diff
@@ -137,6 +139,20 @@ def sales_logs
+  end
+
+  def email_sales_csv
+    codes_only_export = params.require(:codes_only) == "true"
```
ditto

natdeanlewissoftwire replied:
(and i guess in lettings but might not be worth the bother if that's getting tech debted out later anyway)

arfacamble replied:
first bit done, second comment more specific please.

natdeanlewissoftwire replied:
oh actually ignore the second bit, thought this was sales specific but i see it isn't

natdeanlewissoftwire commented on app/helpers/logs_helper.rb:
```diff
@@ -36,6 +36,15 @@ def search_label_for_controller(controller)
+  end
+
+  def csv_download_url_by_log_type(log_type, organisation, search:, codes_only:)
+    if log_type == :lettings
```
teeny tiny: could use a switch like no lines 37-40 to make this ever so slightly more explicit? (and more log-tyoe-symmetrical 😍)

natdeanlewissoftwire replied:
couple of other similar places we could use switches too in this PR

arfacamble replied:
where???

natdeanlewissoftwire commented on app/jobs/email_csv_job.rb:
```diff
@@ -5,14 +5,22 @@ class EmailCsvJob < ApplicationJob
+    end
 
-    filename = organisation.present? ? "logs-#{organisation.name}-#{Time.zone.now}.csv" : "logs-#{Time.zone.now}.csv"
+    filename = "#{[log_type, 'logs', organisation&.name, Time.zone.now].compact.join('-')}.csv"
```
nice

arfacamble replied:
yes

arfacamble committed:
fix linter complaints

arfacamble committed:
minor alterations after rebase to account for changes made on other branches

natdeanlewissoftwire commented on app/views/logs/csv_confirmation.html.erb:
```diff
@@ -9,7 +9,8 @@
 
     <p class="govuk-body">
-      <%= govuk_link_to "Return to logs", lettings_logs_path %>
+      <% logs_path = controller.class.name.start_with?("Lettings") ? lettings_logs_path : sales_logs_path %>
```
could go in a view helper

arfacamble replied:
yes

natdeanlewissoftwire replied:
_should_ go in a view helper

arfacamble replied:
why are you still talking about this?

natdeanlewissoftwire commented on spec/jobs/email_csv_job_spec.rb:
```diff
@@ -39,18 +40,44 @@
+        expect(Csv::SalesLogCsvService).to receive(:new).with(export_type: "labels")
+        codes_only = false
+        job.perform(user, nil, {}, nil, nil, codes_only, "sales")
+        expect(Csv::SalesLogCsvService).to receive(:new).with(export_type: "codes")
```
should these be separate tests?

arfacamble replied:
I don't think it really matters. Happy to be convinced otherwise

natdeanlewissoftwire replied:
this is not the hill i choose to die on

natdeanlewissoftwire reviewed (approved):
lgtm pending comments

arfacamble reviewed (commented):
submit replies to Nat.

arfacamble committed:
refactor after code review

arfacamble committed:
tweak fixtures after rebase containing alterations to the factory defaults
